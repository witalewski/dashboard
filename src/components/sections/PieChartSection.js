import React from 'react';
import { PieChart, Pie, Cell, Legend, Tooltip } from 'recharts';
import { chartColors } from '../../global/styleConstants';
import { Section } from '../Section';

const reformatLabels = className => () => {
  const labels = document.querySelectorAll(
    `.${className} .recharts-pie-label-text tspan`
  );
  for (let i = 0; i < labels.length; i++) {
    labels[i].textContent = `${labels[i].textContent}%`;
  }
};

export const PieChartSection = ({ className, data, title }) => {
  const sum = data.reduce((acc, { value }) => acc + value, 0);
  const percentData = data.map(({ value, label }) => ({
    name: label,
    value: Math.round((value * 100) / sum),
  }));
  return (
    <Section title={title} className={className}>
      <PieChart width={350} height={250}>
        <Pie
          data={percentData}
          cx={120}
          cy={120}
          innerRadius={40}
          outerRadius={80}
          dataKey="value"
          label
          onAnimationEnd={reformatLabels(className)}
        >
          {percentData.map(({ value }, i) => (
            <Cell key={value} fill={chartColors[i % chartColors.length]} />
          ))}
        </Pie>
        <Tooltip />
        <Legend
          width={200}
          align="right"
          layout="vertical"
          wrapperStyle={{ top: 80, left: 260 }}
        />
      </PieChart>
    </Section>
  );
};
