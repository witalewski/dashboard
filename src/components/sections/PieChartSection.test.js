import React from 'react';
import ShallowRenderer from 'react-test-renderer/shallow';
import mockChartData from '../../fixtures/mockChartData.json';
import { PieChartSection } from './PieChartSection';

describe('PieChartSection', () => {
  it('matches snapshot with mock devices data', () => {
    const renderer = new ShallowRenderer();
    renderer.render(
      <PieChartSection title="Devices" data={mockChartData.devices.data} />
    );
    const result = renderer.getRenderOutput();
    expect(result).toMatchSnapshot();
  });

  it('matches snapshot with mock most popular options data', () => {
    const renderer = new ShallowRenderer();
    renderer.render(
      <PieChartSection
        title="Most Popular"
        data={mockChartData.mostPopular.data}
      />
    );
    const result = renderer.getRenderOutput();
    expect(result).toMatchSnapshot();
  });

  it('matches snapshot with mock payment methods data', () => {
    const renderer = new ShallowRenderer();
    renderer.render(
      <PieChartSection
        title="Payment methods"
        data={mockChartData.paymentMethods.data}
      />
    );
    const result = renderer.getRenderOutput();
    expect(result).toMatchSnapshot();
  });
});
