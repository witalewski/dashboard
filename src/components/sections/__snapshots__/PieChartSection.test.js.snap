// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PieChartSection matches snapshot with mock devices data 1`] = `
<Section
  title="Devices"
>
  <PieChart
    barCategoryGap="10%"
    barGap={4}
    cx="50%"
    cy="50%"
    endAngle={360}
    height={250}
    innerRadius={0}
    layout="centric"
    margin={
      Object {
        "bottom": 5,
        "left": 5,
        "right": 5,
        "top": 5,
      }
    }
    outerRadius="80%"
    reverseStackOrder={false}
    stackOffset="none"
    startAngle={0}
    width={350}
  >
    <Pie
      animationBegin={400}
      animationDuration={1500}
      animationEasing="ease"
      blendStroke={false}
      cx={120}
      cy={120}
      data={
        Array [
          Object {
            "name": "Desktop",
            "value": 65,
          },
          Object {
            "name": "Tablet",
            "value": 17,
          },
          Object {
            "name": "Smartphone",
            "value": 14,
          },
          Object {
            "name": "TV",
            "value": 4,
          },
        ]
      }
      dataKey="value"
      endAngle={360}
      fill="#8884d8"
      hide={false}
      innerRadius={40}
      isAnimationActive={true}
      label={true}
      labelLine={true}
      legendType="rect"
      minAngle={0}
      nameKey="name"
      onAnimationEnd={[Function]}
      outerRadius={80}
      paddingAngle={0}
      startAngle={0}
      stroke="#fff"
    >
      <Cell
        fill="#5db990"
      />
      <Cell
        fill="#558ca0"
      />
      <Cell
        fill="#864ba5"
      />
      <Cell
        fill="#829ac0"
      />
    </Pie>
    <Tooltip
      active={false}
      animationDuration={400}
      animationEasing="ease"
      contentStyle={Object {}}
      coordinate={
        Object {
          "x": 0,
          "y": 0,
        }
      }
      cursor={true}
      cursorStyle={Object {}}
      filterNull={true}
      isAnimationActive={true}
      itemSorter={[Function]}
      itemStyle={Object {}}
      labelStyle={Object {}}
      offset={10}
      separator=" : "
      useTranslate3d={false}
      viewBox={
        Object {
          "x1": 0,
          "x2": 0,
          "y1": 0,
          "y2": 0,
        }
      }
      wrapperStyle={Object {}}
    />
    <Legend
      align="right"
      iconSize={14}
      layout="vertical"
      verticalAlign="bottom"
      width={200}
      wrapperStyle={
        Object {
          "left": 260,
          "top": 80,
        }
      }
    />
  </PieChart>
</Section>
`;

exports[`PieChartSection matches snapshot with mock most popular options data 1`] = `
<Section
  title="Most Popular"
>
  <PieChart
    barCategoryGap="10%"
    barGap={4}
    cx="50%"
    cy="50%"
    endAngle={360}
    height={250}
    innerRadius={0}
    layout="centric"
    margin={
      Object {
        "bottom": 5,
        "left": 5,
        "right": 5,
        "top": 5,
      }
    }
    outerRadius="80%"
    reverseStackOrder={false}
    stackOffset="none"
    startAngle={0}
    width={350}
  >
    <Pie
      animationBegin={400}
      animationDuration={1500}
      animationEasing="ease"
      blendStroke={false}
      cx={120}
      cy={120}
      data={
        Array [
          Object {
            "name": "E946347906_AU",
            "value": 36,
          },
          Object {
            "name": "E643516498_AU",
            "value": 4,
          },
          Object {
            "name": "E657734674_AU",
            "value": 17,
          },
          Object {
            "name": "E135478956_AU",
            "value": 44,
          },
        ]
      }
      dataKey="value"
      endAngle={360}
      fill="#8884d8"
      hide={false}
      innerRadius={40}
      isAnimationActive={true}
      label={true}
      labelLine={true}
      legendType="rect"
      minAngle={0}
      nameKey="name"
      onAnimationEnd={[Function]}
      outerRadius={80}
      paddingAngle={0}
      startAngle={0}
      stroke="#fff"
    >
      <Cell
        fill="#5db990"
      />
      <Cell
        fill="#558ca0"
      />
      <Cell
        fill="#864ba5"
      />
      <Cell
        fill="#829ac0"
      />
    </Pie>
    <Tooltip
      active={false}
      animationDuration={400}
      animationEasing="ease"
      contentStyle={Object {}}
      coordinate={
        Object {
          "x": 0,
          "y": 0,
        }
      }
      cursor={true}
      cursorStyle={Object {}}
      filterNull={true}
      isAnimationActive={true}
      itemSorter={[Function]}
      itemStyle={Object {}}
      labelStyle={Object {}}
      offset={10}
      separator=" : "
      useTranslate3d={false}
      viewBox={
        Object {
          "x1": 0,
          "x2": 0,
          "y1": 0,
          "y2": 0,
        }
      }
      wrapperStyle={Object {}}
    />
    <Legend
      align="right"
      iconSize={14}
      layout="vertical"
      verticalAlign="bottom"
      width={200}
      wrapperStyle={
        Object {
          "left": 260,
          "top": 80,
        }
      }
    />
  </PieChart>
</Section>
`;

exports[`PieChartSection matches snapshot with mock payment methods data 1`] = `
<Section
  title="Payment methods"
>
  <PieChart
    barCategoryGap="10%"
    barGap={4}
    cx="50%"
    cy="50%"
    endAngle={360}
    height={250}
    innerRadius={0}
    layout="centric"
    margin={
      Object {
        "bottom": 5,
        "left": 5,
        "right": 5,
        "top": 5,
      }
    }
    outerRadius="80%"
    reverseStackOrder={false}
    stackOffset="none"
    startAngle={0}
    width={350}
  >
    <Pie
      animationBegin={400}
      animationDuration={1500}
      animationEasing="ease"
      blendStroke={false}
      cx={120}
      cy={120}
      data={
        Array [
          Object {
            "name": "VISA",
            "value": 43,
          },
          Object {
            "name": "MasterCard",
            "value": 28,
          },
          Object {
            "name": "PayPal",
            "value": 29,
          },
          Object {
            "name": "iDeal",
            "value": 0,
          },
        ]
      }
      dataKey="value"
      endAngle={360}
      fill="#8884d8"
      hide={false}
      innerRadius={40}
      isAnimationActive={true}
      label={true}
      labelLine={true}
      legendType="rect"
      minAngle={0}
      nameKey="name"
      onAnimationEnd={[Function]}
      outerRadius={80}
      paddingAngle={0}
      startAngle={0}
      stroke="#fff"
    >
      <Cell
        fill="#5db990"
      />
      <Cell
        fill="#558ca0"
      />
      <Cell
        fill="#864ba5"
      />
      <Cell
        fill="#829ac0"
      />
    </Pie>
    <Tooltip
      active={false}
      animationDuration={400}
      animationEasing="ease"
      contentStyle={Object {}}
      coordinate={
        Object {
          "x": 0,
          "y": 0,
        }
      }
      cursor={true}
      cursorStyle={Object {}}
      filterNull={true}
      isAnimationActive={true}
      itemSorter={[Function]}
      itemStyle={Object {}}
      labelStyle={Object {}}
      offset={10}
      separator=" : "
      useTranslate3d={false}
      viewBox={
        Object {
          "x1": 0,
          "x2": 0,
          "y1": 0,
          "y2": 0,
        }
      }
      wrapperStyle={Object {}}
    />
    <Legend
      align="right"
      iconSize={14}
      layout="vertical"
      verticalAlign="bottom"
      width={200}
      wrapperStyle={
        Object {
          "left": 260,
          "top": 80,
        }
      }
    />
  </PieChart>
</Section>
`;
